prov/gni/src/gnix_vc.c

```diff
@@ -96,143 +184,88 @@ static inline void __gnix_vc_set_ht_key(void *gnix_addr,
-
-	vc = (struct gnix_vc *)_gnix_ht_lookup(ep->vc_ht, key);
-
 	if (vc == NULL) {
-		ret = _gnix_vc_alloc(ep,
-				     av_entry,
-				     &vc_tmp);
+		/* Look up full AV entry for the destination address. */
+		ret = _gnix_av_lookup(av, dest_addr, &av_entry);
+		if (ret != FI_SUCCESS) {
+			GNIX_WARN(FI_LOG_EP_DATA,
+				  "_gnix_av_lookup for addr 0x%llx returned %s \n",
+				  dest_addr, fi_strerror(-ret));
+			goto err_w_lock;
+		}
+
+		/* Initiate a connection to the endpoint. */
+		ret = _gnix_vc_alloc(ep, av_entry, &vc_tmp);
 		if (ret != FI_SUCCESS) {
 			GNIX_WARN(FI_LOG_EP_DATA,
 				  "_gnix_vc_alloc returned %s\n",
 				  fi_strerror(-ret));
 			goto err_w_lock;
 		}
 
-		ret = _gnix_ht_insert(ep->vc_ht, key,
-				      vc_tmp);
-		if (likely(ret == FI_SUCCESS)) {
-			vc = vc_tmp;
-			vc->modes |= GNIX_VC_MODE_IN_HT;
-			/*
-			 * if FI_AV_TABLE is being used, also insert
-			 * in to vector lookup table
-			 */
-			if (av_type == FI_AV_TABLE) {
-				ret = _gnix_vec_insert_at(ep->vc_table, (void *)vc_tmp,
-							  dest_addr);
-				if (ret != FI_SUCCESS) {
-					GNIX_WARN(FI_LOG_EP_DATA,
-						  "_gnix_vec_insert_at returned %s\n",
-						  fi_strerror(-ret));
-				} else {
-					vc_tmp->modes |= GNIX_VC_MODE_IN_TABLE;
-				}
-			}
-			fastlock_release(&ep->vc_lock);
-			ret = _gnix_vc_connect(vc);
-			if (ret != FI_SUCCESS) {
-				GNIX_WARN(FI_LOG_EP_DATA,
-					  "_gnix_vc_connect returned %s\n",
-					  fi_strerror(-ret));
-				goto err;
-			}
-		} else if (ret == -FI_ENOSPC) {
-			_gnix_vc_destroy(vc_tmp);
-			vc = _gnix_ht_lookup(ep->vc_ht, key);
-			fastlock_release(&ep->vc_lock);
-			assert(vc != NULL);
-			assert(vc->modes & GNIX_VC_MODE_IN_HT);
-			ret = FI_SUCCESS;
-		} else {
+		/* Store new VC in EP connection table. */
+		ret = _gnix_ep_vc_store(ep, vc_tmp, dest_addr);
+		if (unlikely(ret != FI_SUCCESS)) {
 			GNIX_WARN(FI_LOG_EP_DATA,
-				  "_gnix_ht_insert returned %s\n",
+				  "_gnix_ep_vc_store returned %s\n",
 				  fi_strerror(-ret));
 			goto err_w_lock;
 		}
-	} else  {
 
-		/*
-		 * check here if the vc should be in the table accelerated
-		 * lookup but hasn't yet been inserted
-		 */
+		vc = vc_tmp;
+		vc->modes |= GNIX_VC_MODE_IN_HT;
 
-		if ((av_type == FI_AV_TABLE) &&
-			!(vc->modes & GNIX_VC_MODE_IN_TABLE)) {
-			ret = _gnix_vec_insert_at(ep->vc_table, (void *)vc,
-						  dest_addr);
-			if (ret != FI_SUCCESS) {
-				GNIX_WARN(FI_LOG_EP_DATA,
-					  "_gnix_vec_insert_at returned %s\n",
-					  fi_strerror(-ret));
-			} else {
-				vc->modes |= GNIX_VC_MODE_IN_TABLE;
-			}
+		ret = _gnix_vc_connect(vc);
+		if (ret != FI_SUCCESS) {
+			GNIX_WARN(FI_LOG_EP_DATA,
+				  "_gnix_vc_connect returned %s\n",
+				  fi_strerror(-ret));
+			goto err_w_lock;
 		}
-
-		fastlock_release(&ep->vc_lock);
 	}
 
+	COND_RELEASE(ep->requires_lock, &ep->vc_lock);
+
 	*vc_ptr = vc;
 	return ret;
 
 err_w_lock:
-	fastlock_release(&ep->vc_lock);
-err:
+	COND_RELEASE(ep->requires_lock, &ep->vc_lock);
 	if (vc != NULL)
```

Comment:
I don't think that's necessary right now.  If we find an unmapped VC we only unlock and return (skipping the whole 'if (vc == NULL) block).  There's no other way to fail.